### PR TITLE
Making video source external

### LINKED ISSUE #57: Make video source external
Extracting the source of the video frames out of `VisionManager` will help in the following manner:
- user is able to control the settings of the camera,
- user is able to obtain and utilize the same frames `VisionManager` is using,
- user is free to implement any source of video frames (e.g. external wi-fi camera, movie file, etc.) and provide it to the `VisionManager`,
- `VisionManager` doesn't make any assumptions whether video stream should start or stop, it just subscribes to it

### PR DESCRIPTION
**Pull request by chezzdev**
Resolves #57

### CODE REVIEW

File: MapboxVision/Services/VideoSource/CameraVideoSource.swift [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+            if let cameraParameters = getCameraParameters(sampleBuffer: sampleBuffer) {
+                observer.videoSource(self, didOutput: cameraParameters)
+            }
+        }
+    }
+    
+    public func captureOutput(_ output: AVCaptureOutput, didDrop sampleBuffer: CMSampleBuffer, from connection: AVCaptureConnection) {
+        var mode: CMAttachmentMode = 0
+        guard let reason = CMGetAttachment(sampleBuffer, kCMSampleBufferAttachmentKey_DroppedFrameReason, &mode) else { return }
+        print("Sample buffer was dropped. Reason: \(reason)")
```
Comment (yunikkk) [inline MapboxVision/Services/VideoSource/CameraVideoSource.swift]: Is print needed or added just for debug purposes?

File: MapboxVision/VideoStream/VisionManager.swift [hunk 2]
```diff
@@ -710,6 +696,35 @@ public final class VisionManager {
...
+        guard state.isStarted else { return }
+        
+        dependencies.recorder.handleFrame(videoSample.buffer)
+        
+        dependencies.core.setImage(pixelBuffer)
+    }
+    
+    public func videoSource(_ videoSource: VideoSource, didOutput cameraParameters: CameraParameters) {
+        // TODO: use new camera params
+        dependencies.core.setCameraWidth(
```
Comment (yunikkk) [inline MapboxVision/VideoStream/VisionManager.swift]: Is method named `setCameraWidth`?
Also, are `-1` values really intended here?

File: MapboxVision/Services/VideoSource/CameraVideoSource.swift (hunk 1 above)
Comment (chezzdev) [reply MapboxVision/Services/VideoSource/CameraVideoSource.swift]: `print` is our current logging system.

File: MapboxVision/VideoStream/VisionManager.swift (hunk 2 above)
Comment (chezzdev) [reply MapboxVision/VideoStream/VisionManager.swift]: There's no fields for focal length in pixels in current core API. It'll change in a new version and the correct method will be utilized here.

File: MapboxVision/Services/VideoSource/CameraVideoSource.swift [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    
+    open var isExternal: Bool {
+        return false
+    }
+}
+
+extension CameraVideoSource: AVCaptureVideoDataOutputSampleBufferDelegate {
+    public func captureOutput(_ output: AVCaptureOutput, didOutput sampleBuffer: CMSampleBuffer, from connection: AVCaptureConnection) {
+        notify { (observer) in
+            let sample = VideoSample(buffer: sampleBuffer, format: .bgra)
```
Comment (vanmaxim) [inline MapboxVision/Services/VideoSource/CameraVideoSource.swift]: Transform 'format' to constant. Implement helper for cvPixelFormat

File: MapboxVision/Services/VideoSource/VideoSource.swift [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    public let format: InputImageFormat
+    
+    public init(buffer: CMSampleBuffer, format: InputImageFormat) {
+        self.buffer = buffer
+        self.format = format
+    }
+}
+
+public protocol VideoSource: class {
+    typealias VideoSampleOutput = (VideoSample) -> Void
```
Comment (vanmaxim) [inline MapboxVision/Services/VideoSource/VideoSource.swift]: plz remote typealiases

**Comment by chezzdev** [comment]
@vanmaxim @yunikkk PTAL

**Pull request merged**